core-backend/src/funcs.rs

```diff
@@ -183,26 +151,55 @@ where
 macro_rules! impl_syscall_builder {
     ($($generic:ident),+) => {
         #[allow(non_snake_case)]
-        impl<Caller, Res, Call, Builder, $($generic),+> SyscallBuilder<Caller, ($($generic,)+), Res, Call>
+        impl<Caller, Ext, Res, Call, Builder, $($generic),+> SyscallBuilder<Caller, ($($generic,)+), Res, Call>
             for Builder
         where
             Builder: FnOnce($($generic),+) -> Call,
             Call: Syscall<Caller, Res>,
-            $( $generic: TryFrom<SyscallValue, Error = HostError>,)+
+            Caller: AsContextExt<State = HostState<Ext, BackendMemory<ExecutorMemory>>>,
+            Ext: BackendExternalities + 'static,
+            $( $generic: SyscallArg, )+
         {
-            fn build(self, args: &[Value]) -> Result<Call, HostError> {
-                const ARGS_AMOUNT: usize = impl_syscall_builder!(@count $($generic),+);
+            fn build(self, caller: &mut CallerWrap<Caller>, args: &[Value]) -> Result<Call, HostError> {
+                let ARGS_AMOUNT: usize = 0 $( + $generic::REQUIRED_ARGS )+;
+
+                if args.len() != ARGS_AMOUNT {
+                    return Err(HostError);
+                }
 
-                let [$($generic),+]: [Value; ARGS_AMOUNT] = args.try_into().map_err(|_| HostError)?;
+                let mut registry: Option<MemoryAccessRegistry<Caller>> = None;
+
+                $(
+                    if $generic::REQUIRES_MEMORY_MANAGER && registry.is_none() {
+                        registry = Some(MemoryAccessRegistry::default());
+                    }
+                )+
+
+                let mut index = 0;
                 $(
-                    let $generic = SyscallValue($generic).try_into()?;
+                    let args_count = $generic::REQUIRED_ARGS;
+                    let args_slice = &args[index..index + args_count];
+                    let $generic = $generic::pre_process(&mut registry, args_slice).map_err(|_| HostError)?;
+                    index += args_count;
                 )+
-                Ok((self)($($generic),+))
+
+                if(index != ARGS_AMOUNT) {
+                    return Err(HostError);
+                }
+
+                if registry.is_some() {
+                    let io = registry.unwrap().pre_process(caller);
```

Comment:
Error can be set like in `CallerWrap::set_termination_reason()` and `HostError` returned, so you don't need to handle late return

Reply:
This would be a great idea, we could get rid of a lot of code, except for the fact that when an error occurs in FallibleSyscall, the error content is written to memory in different formats depending on the type of syscall